persistent/Database/Persist/Types/Base.hs

```diff
@@ -297,24 +297,40 @@ data EmbedFieldDef = EmbedFieldDef
     }
     deriving (Show, Eq, Read, Ord, Lift)
 
+-- | Returns 'True' if the 'FieldDef' does not have a 'MigrationOnly' or
+-- 'SafeToRemove' flag from the QuasiQuoter.
+--
+-- @since 2.13.0.0
+isHaskellField :: FieldDef -> Bool
+isHaskellField fd =
+    FieldAttrMigrationOnly `notElem` fieldAttrs fd &&
+    FieldAttrSafeToRemove `notElem` fieldAttrs fd
+
 toEmbedEntityDef :: EntityDef -> EmbedEntityDef
 toEmbedEntityDef ent = embDef
   where
     embDef = EmbedEntityDef
-      { embeddedHaskell = entityHaskell ent
-      , embeddedFields = map toEmbedFieldDef $ entityFields ent
-      }
+        { embeddedHaskell = entityHaskell ent
+        , embeddedFields =
+            map toEmbedFieldDef
+            $ filter isHaskellField
```

Comment:
If we're embedding the entity, then we only care about the Haskell variants. We can't have a normal SQL schema, so it's inherently flexible/untyped.